File: Problems & Solutions/907. Sum of Subarray Minimums/25 November 2022/Approach 1 - Monotonic Stack - Contribution of Each Element/Sum of Subarray Minimums.java
class Solution {
    public int sumSubarrayMins(int[] arr) {
        int MOD = 1000000007;

        Stack<Integer> stack = new Stack<>();
        long sumOfMinimums = 0;

        for (int i = 0; i <= arr.length; i++) {

            while (!stack.empty() && (i == arr.length || arr[stack.peek()] >= arr[i])) {

                int mid = stack.pop();
                int leftBoundary = stack.empty() ? -1 : stack.peek();
                int rightBoundary = i;

                long count = (mid - leftBoundary) * (rightBoundary - mid) % MOD;

                sumOfMinimums += (count * arr[mid]) % MOD;
                sumOfMinimums %= MOD;
            }
            stack.push(i);
        }

        return (int) (sumOfMinimums);
    }
}

File: Problems & Solutions/907. Sum of Subarray Minimums/25 November 2022/Approach 1 - Monotonic Stack - Contribution of Each Element/Sum of Subarray Minimums.py
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        MOD = 10 ** 9 + 7
        stack = []
        sum_of_minimums = 0;

        for i in range(len(arr) + 1):
            
            while stack and (i == len(arr) or arr[stack[-1]] >= arr[i]):

                mid = stack.pop()
                left_boundary = -1 if not stack else stack[-1]
                right_boundary = i

                count = (mid - left_boundary) * (right_boundary - mid)
                sum_of_minimums += (count * arr[mid])

            stack.append(i)
        
        return sum_of_minimums % MOD
